### PR TITLE
Tf slim for review

### PR DESCRIPTION
**Pull request by cramraj8**

### CODE REVIEW

File: deep_regression.py [hunk 1]
```diff
@@ -0,0 +1,251 @@
+from __future__ import absolute_import, division, print_function
+
```
Comment (safooray) [inline deep_regression.py]: Remove extra line.

File: deep_regression.py [hunk 2]
```diff
@@ -0,0 +1,251 @@
...
+import tensorflow as tf
+import matplotlib.pyplot as plt
+import numpy as np
+import scipy.io as sio
+import pandas as pd
+from sklearn.utils import shuffle
+from sklearn.model_selection import train_test_split
+import tensorflow.contrib.slim as slim
+from tensorflow.contrib.framework.python.ops.variables \
```
Comment (safooray) [inline deep_regression.py]: It's ok to have import lines longer than 80 chars. As far as I know you are advised not to use \ for line continuation.

File: deep_regression.py [hunk 3]
```diff
@@ -0,0 +1,251 @@
...
+    import get_or_create_global_step
+from tensorflow.python.platform import tf_logging as logging
+
+
+'''
+Changeable parameters OR variables
```
Comment (safooray) [inline deep_regression.py]: We don't need this comment (lines 17-36). If you want the user to be able to change these parameters, make this a class and set these in the constructor. Alternatively, you can pass them as flags.

File: deep_regression.py [hunk 4]
```diff
@@ -0,0 +1,251 @@
...
+7. Hidden layer neurons number
+    N_HIDDEN_1
+    N_HIDDEN_2
+    N_HIDDEN_3
+    N_HIDDEN_4
+8. VARIANCE (for weights & biases, random normal initialization)
+9. tf.reduce_sum OR tf.reduce_mean
+10. Regularization method
+11. Gradient Descend Method
+    optimizer = tf.train.GradientDescentOptimizer()  ===> Gives all NaN values
```
Comment (safooray) [inline deep_regression.py]: Then something is wrong and should be resolved. Maybe learning rate is too big?

File: deep_regression.py [hunk 5]
```diff
@@ -0,0 +1,251 @@
...
+NUM_OF_EPOCHS_BEFORE_DECAY = 1000
+
+# ============ Network Parameters ============
+HIDDEN_LAYERS = [500, 400, 600, 500, 500, 500]
+N_CLASSES = 1                       # Only one output prediction column
+
+
+def load_data_set(name=None):
+    """
+    This function reads the data file and extracts the features and labelled
```
Comment (safooray) [inline deep_regression.py]: Description does not make sense. Make it clear and concise.

File: deep_regression.py [hunk 6]
```diff
@@ -0,0 +1,251 @@
...
+    values.
+    Then according to that patient is dead, only those observations would be
+    taken care
+    for deep learning trainig.
+  Args:
+    Nothing
```
Comment (safooray) [inline deep_regression.py]: name?

File: deep_regression.py [hunk 7]
```diff
@@ -0,0 +1,251 @@
...
+  Returns:
+    `Numpy array`, extracted feature columns and label column.
+  Example:
+    >>> read_dataset()
+    ( [[2.3, 2.4, 6.5],[2.3, 5.4,3.3]], [12, 82] )
+    """
+    data_feed = pd.read_csv('Brain_Integ_X.csv', skiprows=[0], header=None)
```
Comment (safooray) [inline deep_regression.py]: Dataset names should not be hard coded. Pass in as argument.

File: deep_regression.py [hunk 8]
```diff
@@ -0,0 +1,251 @@
...
+    labels_feed = pd.read_csv('Brain_Integ_Y.csv', skiprows=[1], header=0)
+    survival = labels_feed['Survival']
+    censored = labels_feed['Censored']
+
+    survival = survival.values
+    censored = censored.values
+    data = data_feed.values
+    data = np.float32(data)
+
+    censored_survival = survival[censored == 1]
```
Comment (safooray) [inline deep_regression.py]: Why are you keeping censored patients? Censored means alive.

File: deep_regression.py [hunk 9]
```diff
@@ -0,0 +1,251 @@
...
+
+    y = np.asarray(censored_survival)
+    x = np.asarray(censored_data)
+
+    print('Shape of X : ', x.shape)
+    print('Shape of Y : ', y.shape)
+    return (x, y)
+
+
+# ******************************************************************************
```
Comment (safooray) [inline deep_regression.py]: Two empty lines and nothing more.

File: deep_regression.py [hunk 10]
```diff
@@ -0,0 +1,251 @@
...
+
+
+with tf.Graph().as_default() as graph:
```
Comment (safooray) [inline deep_regression.py]: This should be in a module named "train", while load_data should be in a module named "data_providers" or something.

File: deep_regression.py [hunk 11]
```diff
@@ -0,0 +1,251 @@
...
+    data_x, data_y = load_data_set()
+    data_x, data_y = shuffle(data_x, data_y, random_state=1)
+    x_train, x_test, y_train, y_test = train_test_split(data_x, data_y,
+                                                        test_size=0.20,
+                                                        random_state=420)
+
+    total_observations = x_train.shape[0]         # Total Samples
+    input_features = x_train.shape[1]             # number of columns(features)
+
+    log_dir = './log/'
```
Comment (safooray) [inline deep_regression.py]: Should not be hard-coded. Pass in as flag.
Here and elsewhere.

File: deep_regression.py [hunk 12]
```diff
@@ -0,0 +1,251 @@
...
+with tf.Graph().as_default() as graph:
+    logging.set_verbosity(tf.logging.INFO)  # Set the verbosity to INFO level
+
+    data_x, data_y = load_data_set()
+    data_x, data_y = shuffle(data_x, data_y, random_state=1)
+    x_train, x_test, y_train, y_test = train_test_split(data_x, data_y,
+                                                        test_size=0.20,
+                                                        random_state=420)
+
+    total_observations = x_train.shape[0]         # Total Samples
```
Comment (safooray) [inline deep_regression.py]: Do tuple expansion in 1 line instead of 2.

File: deep_regression.py [hunk 13]
```diff
@@ -0,0 +1,251 @@
...
+# ******************************************************************************
+
+
+with tf.Graph().as_default() as graph:
+    logging.set_verbosity(tf.logging.INFO)  # Set the verbosity to INFO level
+
+    data_x, data_y = load_data_set()
+    data_x, data_y = shuffle(data_x, data_y, random_state=1)
+    x_train, x_test, y_train, y_test = train_test_split(data_x, data_y,
+                                                        test_size=0.20,
```
Comment (safooray) [inline deep_regression.py]: These numbers should either be constants or flags.

File: deep_regression.py [hunk 14]
```diff
@@ -0,0 +1,251 @@
...
+    if not tf.gfile.Exists(ckpt_dir):
+        tf.gfile.MakeDirs(ckpt_dir)
+
+    num_batches_per_epoch = total_observations / BATCH_SIZE
+    num_steps_per_epoch = num_batches_per_epoch
+    decay_steps = int(NUM_OF_EPOCHS_BEFORE_DECAY * num_steps_per_epoch)
+
+    global_step = tf.Variable(0, name='global_step', trainable=False)
+
+    # Create the model inference
```
Comment (safooray) [inline deep_regression.py]: what is "model inference"?

File: deep_regression.py [hunk 15]
```diff
@@ -0,0 +1,251 @@
...
+    def multilayer_neural_network_model(inputs, HIDDEN_LAYERS, BETA,
```
Comment (safooray) [inline deep_regression.py]: input arguments must be lower case.

File: deep_regression.py [hunk 16]
```diff
@@ -0,0 +1,251 @@
...
+                                        scope="deep_regression_model"):
+        """Creates a deep regression model.
+
+        Args:
```
Comment (safooray) [inline deep_regression.py]: Missing arguments.

File: deep_regression.py [hunk 17]
```diff
@@ -0,0 +1,251 @@
...
+            inputs: A node that yields a `Tensor` of size [total_observations,
+            input_features].
+
+        Returns:
+            predictions: `Tensor` of shape [1] (scalar) of response.
```
Comment (safooray) [inline deep_regression.py]: shape should be in (), not in []

File: deep_regression.py [hunk 18]
```diff
@@ -0,0 +1,251 @@
...
+            end_points = {}
+            with slim.arg_scope([slim.fully_connected],
+                                activation_fn=tf.nn.relu,
+                                weights_regularizer=slim.l2_regularizer(BETA)):
+                net = slim.stack(inputs,
+                                 slim.fully_connected,
+                                 HIDDEN_LAYERS,
+                                 scope='fc')
+                end_points['fc'] = net
+                predictions = slim.fully_connected(net, 1, activation_fn=None,
```
Comment (safooray) [inline deep_regression.py]: Use the constant N_CLASSES here instead of hard coding 1.

File: deep_regression.py [hunk 19]
```diff
@@ -0,0 +1,251 @@
...
+
+    # tf.Graph input
+    x = tf.placeholder("float", [None, input_features], name='features')
+    y = tf.placeholder("float", [None], name='labels')
+
+    # Construct the grpah for forward pass
+    pred, end_points = multilayer_neural_network_model(x, HIDDEN_LAYERS, BETA)
+
+    # Print name and shape of each tensor.
+    print("Layers")
```
Comment (safooray) [inline deep_regression.py]: Remove unnecessary prints.

File: deep_regression.py [hunk 20]
```diff
@@ -0,0 +1,251 @@
...
+
+    ckpt_dir = './regression_model/'
+    if not tf.gfile.Exists(ckpt_dir):
+        tf.gfile.MakeDirs(ckpt_dir)
+
+    num_batches_per_epoch = total_observations / BATCH_SIZE
+    num_steps_per_epoch = num_batches_per_epoch
+    decay_steps = int(NUM_OF_EPOCHS_BEFORE_DECAY * num_steps_per_epoch)
+
+    global_step = tf.Variable(0, name='global_step', trainable=False)
```
Comment (safooray) [inline deep_regression.py]: This shouldn't be necessary given that you are using get_or_create later

File: deep_regression.py [hunk 21]
```diff
@@ -0,0 +1,251 @@
...
+
+    # Define your exponentially decaying learning rate
+    lr = tf.train.exponential_decay(learning_rate=INITIAL_LEARNING_RATE,
+                                    global_step=global_step,
+                                    decay_steps=decay_steps,
+                                    decay_rate=LEARNING_RATE_DECAY_FACTOR,
+                                    staircase=True)
+
+# ********************** LOSS && OPTIMIZE *************************************
+
```
Comment (safooray) [inline deep_regression.py]: Remove empty line.

File: deep_regression.py [hunk 22]
```diff
@@ -0,0 +1,251 @@
...
+    mean_squared_error_loss = tf.reduce_sum(
+        tf.square(tf.transpose(pred) - y),
+        name="loss") / (2 * total_observations)
+
+    regularization_loss = tf.add_n(tf.losses.get_regularization_losses())
+
+    total_loss = mean_squared_error_loss + regularization_loss
+    # total_loss = tf.losses.get_total_loss(add_regularization_losses=True)
+
+#                   ************************************
```
Comment (safooray) [inline deep_regression.py]: Remove unnecessary comments and empty lines. Everywhere.

File: deep_regression.py [hunk 23]
```diff
@@ -0,0 +1,251 @@
...
+
+    optimizer = tf.train.AdamOptimizer(learning_rate=lr).\
```
Comment (safooray) [inline deep_regression.py]: Do not use \ for line continuation.

File: deep_regression.py [hunk 24]
```diff
@@ -0,0 +1,251 @@
...
+    with tf.Session() as sess:
+
+        sess.run(tf.global_variables_initializer())
+
+        # Create a saver object which will save all the variables
+        saver = tf.train.Saver()
+
+        for epoch in range(TRAINING_EPOCHS + 1):
+            avg_cost = 0.0
+            total_batch = int(total_observations / BATCH_SIZE)
```
Comment (safooray) [inline deep_regression.py]: 1. 'total_batch' does not make sense. n_batches?
2. Why don't you reuse your num_batches_per_epoch?

File: deep_regression.py [hunk 25]
```diff
@@ -0,0 +1,251 @@
...
+            for i in range(total_batch - 1):    # Loop over all batches
```
Comment (safooray) [inline deep_regression.py]: use tf.train.shuffle_batch, so you don't have to do shuffling and batching manually, and reduce your code size.

File: deep_regression.py [hunk 26]
```diff
@@ -0,0 +1,251 @@
...
+                batch_x = x_train[i * BATCH_SIZE:(i + 1) * BATCH_SIZE]
+                batch_y = y_train[i * BATCH_SIZE:(i + 1) * BATCH_SIZE]
+                _, batch_cost, batch_pred = sess.run([optimizer,
```
Comment (safooray) [inline deep_regression.py]: Can you use slim training loop instead?

File: deep_regression.py [hunk 27]
```diff
@@ -0,0 +1,251 @@
...
+        # TESTING PHASE #
+        _, test_predicted_values = sess.run([total_loss, pred],
+                                            feed_dict={x: x_test,
+                                                       y: y_test})
+        for i in range(len(y_test)):
+            print("Labeled VALUE : ", y_test[i], " \t\t\t \
+                >>>> Predicted VALUE : ", float(test_predicted_values[i]))
+
+
+#### ------------------ PLOTTING -------------------------------------- ####
```
Comment (safooray) [inline deep_regression.py]: Move plotting to separate module.

File: deep_regression.py [hunk 28]
```diff
@@ -0,0 +1,251 @@
...
+                      " cost = ", " {:.9f} ".format(avg_cost))
+                for i in xrange(3):  # Comparing orig, predicted values
+                    print("label value:", batch_y[i], "predicted value:",
+                          batch_pred[i])
+                print("------------------------------------------------------")
+        print('Global_Step final value : ', sess.run(global_step))
+        # Saving the Graph and Variables in Checkpoint files
+        saver.save(sess, ckpt_dir + 'deep_regression_trained_model',
+                   global_step=global_step)
+        pred_out = sess.run(pred, feed_dict={x: x_train})
```
Comment (safooray) [inline deep_regression.py]: 'pred_value' is a better name.

Comment (safooray) [inline deep_regression.py]: Did you mean to feed in the test set this time? Because you have already evaluated pred on train set batch by batch.

**Review by safooray — COMMENTED** [review]
I left several comments. A lot of them are regarding readability, style, cleanness, etc. Once you get these out of the way we can actually look at the bigger picture and work on the methods.

File: deep_regression.py (hunk 5 above)
Comment (cramraj8) [reply deep_regression.py]: I read somewhere like, in the docstring, the core job of the function should be expressed in the first line. And multiple lines description of the first line should be given on the following lines. Is that so ?

File: deep_regression.py (hunk 23 above)
Comment (cramraj8) [reply deep_regression.py]: So we can split one expression into two, to make that fit into 80 chars?

File: deep_regression.py (hunk 28 above)
Comment (cramraj8) [reply deep_regression.py]: I already evaluated the train prediction batch by batch. But I need whole train dataset prediction values for visualizing. However, rather than appending those batch by batch values into another array, evaluating sess.run() seemed comfortable. Is this a bad style ?

File: deep_regression.py (hunk 8 above)
Comment (cramraj8) [reply deep_regression.py]: When I change C ==0, then number of observations become 160. The training misbehaves very badly. Should I tune parameters in this case?
